### DISCUSSION TITLE
"Headless" az login, serverside

### DISCUSSION THREAD

**Discussion by giuliohome** [Azure/kubelogin#163] [Q&A] [answered]
Is there a way to print/get also the client id or - in other words - is there a reason why it is omitted from the console log?

I mean, I've not yet looked at the code, but I assume it's calling the browser via a complete URL with the client id as a query parameter, so I wonder why it's not logged as well. Before thinking about forking the repo and possibly changing the code, I thought I'd rather start asking here first.

My goal is to run `az login`and `kubectl` on a pure server (hopefully without any X gui) and open the browser on another workstation, eventually coming back to the server.

**Comment by giuliohome** [comment]
I see that the interactive module calls an oauth client that returns the token. I guess the browser can't be opened on a different machine, because the result must be returned from the internal api call...

Otherwise one would follow a different flow (maybe the device code or a non interactive flow). In particular I've noticed that one can perform:
- az login interactive with browser and produce kubeconfig on a machine
- kubecectl with previously produced kubeconfig on a different machine
- authorize the above required device code on another machine

**Comment by giuliohome** [comment, marked as answer]
Btw, this is the az login [prompt](https://github.com/Azure/azure-cli/blob/f0c87ca820e0bd3f9dd403ef96c21ed49b734d16/src/azure-cli-core/azure/cli/core/auth/identity.py#L141) I'm talking about. For the sake of it, could one derive and show the client id at that point? 
Nope, the client id comes from the [access token](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/acebccb5b3b49e96b8b6c910d8c0ff9820a7d18e/apps/internal/oauth/oauth.go#L141) that in turn is obtained after logging in with username and password: that finally explains why it is not in the `az login` log (before the browser is opened and the user is logged)!
From the [README](https://github.com/AzureAD/microsoft-authentication-library-for-go#public-surface):
> If there is no suitable token in the cache, or you choose to skip this step, now we can send a request to AAD to obtain a token.

**Comment by giuliohome** [comment]
hey guys there, 

it looks to me that [Azure AD federated identity credentials](https://docs.microsoft.com/en-us/graph/api/resources/federatedidentitycredentials-overview?view=graph-rest-beta) solve the issue [Enable authentication via environment variables](https://github.com/Azure/azure-cli/issues/10241), don't they? 

(they're also non-interactive, hence "better" from the point of view of my scenario here)

**Comment by weinong** [comment]
if you want to have non-ineteractive login, you should use service principal.

**Reply by giuliohome** [reply]
I know about using a service principal and this is not what I'm talking about. Furthermore I'm trying to study and understand things in depth, I don't need a motto or a leitmotiv here: I'm posting the results of a bit more involved analysis

**Reply by weinong** [reply]
all i'm trying to say is user login like device code flow or interactive web login are not suited for headless/non-interactive flow. if your server is Azure VM, you can also use managed service identity (-l msi). Another solution is ropc, but it's really a bad practice.

**Reply by giuliohome** [reply]
Assuming I have a user login and it is interactive, I'm trying to understand what happens.
For example, can I open the web browser on a machine and login into another (I've tried, yes, the details matter however).

Another question on top of my mind (always about interactive user login). Are interactive web login and device code flow 2 alternatives? But I see both happening in my scenario... first is requested a web login and then kubectl asks for device code...

And so on and so forth...  (Btw, yes our servers are in Azure and we also use non interactive logins, but that was not what I was studying here)

**Reply by weinong** [reply]
> For example, can I open the web browser on a machine and login into another (I've tried, yes, the details matter however).

Yes, you can do that with device code flow. And, yes, it's considered a security compromise. Hence, the new "web interactive login", which is the default used by `az` cli, is considered the replacement for device code flow. With web interactive login, organization can enforce conditional access policy such as restricting login only from organization IP or managed devices. How the web interactive login works is the AAD login will redirect to a localhost webserver

**Reply by giuliohome** [reply]
> For example, can I open the web browser on a machine and login into another (I've tried, yes, the details matter however).

> Yes, you can do that with device code flow. 

I don't think this is what I am doing. Again, I guess the exact details of what I am doing are relevant.
Let's start on a machine A.
When the az login completes, how does it  store and "pass" the login info (or data or config or whatever) to kubectl? 
Are all the needed data contained in a file called kubeconfig?
So, what happens if I move that kubeconfig to machine B and I run kubectl on B?
How is that different than running kubectl on A (in specific tech details, wrt the open source code in the repo)?

**Reply by weinong** [reply]
are you using kubelogin with -l azurecli or -l devicecode?

**Reply by giuliohome** [reply]
I guess with -l azurecli (yes the goal is to use interactive AAD login via kubelogin), but even before that, the instructions I received tell me to do "az login" as first step (and I'm not convinced those instructions are perfect, hence I'm studying it by myself). Just that.

These are the instructions:
1 download kubelogin and put it in the path
2 just do "az login" 
3 az aks get-credentials ... blablabla ... > kubeconfig.yaml
kubectl get pods ...blablabla ...  --kubeconfig kubeconfig.yaml

I'm not even sure the instructions are fully correct, if I compare them to the readme of this repo.

**Reply by weinong** [reply]
when you do `az login`, it defaults to interactive web login as I described earlier. It results in a refresh token and an access token stored in encrypted form on the filesystem. when you use kubelogin with -l azurecli, it basically does `az account get-access-token --resource ${SERVER_APP_ID}` where `${SERVER_APP_ID}` is what is specified in `kubelogin get-tokon --server-id ${SERVER_APP_ID}`. This command will use the refresh token to mint a new access token issued for `${SERVER_APP_ID}`, and kubelogin returns this token to kubectl to send to api-server.

**Reply by giuliohome** [reply]
So, are you saying that those instructions are wrong somehow? (however following those instructions, I can login eventually) Instead of saying "az login" at step 2, should they say instead  kubelogin with -l azurecli?  Btw in your readme, I see " -l azurecli" ony mentioned in the Azure CLI token login (non interactive), which is not what you are writing here: I think you are writing about https://github.com/Azure/kubelogin#web-browser-flow-default and what is this

```
kind: Config
preferences: {}
users:
  - name: user-name
    user:
      exec:
        apiVersion: client.authentication.k8s.io/v1beta1
        args:
        - get-token
        - --login
        - interactive
        - --server-id
        - <AAD server app ID>
        - --client-id
        - <AAD client app ID>
        - --tenant-id
        - <AAD tenant ID>
        - --environment
        - AzurePublicCloud
        command: kubelogin
```
(with no explanation) from the readme? A kubeconfig to be manually prepared? The readme doesn't explain how to get there, how to retrieve the values and prepare that kubeconfig. Or is it the `az login` that ultimately does it? And is it simpler and correct or should we really be using a different kubelogin command to prepare such a kubeconfig? If I look at the kubeconfig generated at the step 3 (following the instructions 1,2,3, I mentioned before) it is very similar to the above! 
well, not exactly, under the tenant-id I see
```
      - --login
      - devicecode
```
So, it looks like we do both az cli before and then also device code (only the first time)!
So the instructions 2 az login and 3 az aks get-credentials ... blablabla ... > kubeconfig.yaml appear to be the correct! (except, maybe, the further step with the device code, is it a further security check?) And again, aside all this, I've also experimented that one can bring the kubeconfig.yaml on another machine (e.g. a pure server without a browser)

**Reply by weinong** [reply]
what doc are you referring to?
this is how I will go about it
```sh
az login
az aks get-credentials ...
kubelogin convert-kubeconfig -l azurecli
kubectl ...
```

**Reply by giuliohome** [reply]
An internal doc, that we use...
Well, so indeed I see you are adding a step with `kubelogin convert-kubeconfig -l azurecli` that is not in our doc.
In fact I was wondering about it and I asked them (should I insist that this step is missing?).
I don't know if our doc is wrong or incomplete. I'm not sure if that `kubelogin convert-kubeconfig -l azurecli` is necessary or what it is exactly doing. I assume you would redirect the previous az aks get-credential into a kubeconfig and then do you need to convert that kubeconfig?

**Reply by weinong** [reply]
AKS will return exec format kubeconfig with `devicecode` login for any 1.24+ clusters.
So, if you want to use `azurecli` login, `kubelogin convert-kubeconfig -l azurecli` will be required.

For any < 1.24 clusters, `kubelogin convert-kubeconfig` is required if you want to use kubelogin
